Improvements!

ccutrer opened the pull request:
- don't bother encrypting the IV (?!)
  - use the same encryption key for the encryption as for the HMAC
    (easy to configure, since it's the same as for normal CookieStore)
  - compress the session before encrypting it (optional, will skip
    compression if it actually makes it larger)
  - use AES 128 CBC (AES 256 is actually _weaker_, due to a known
    attack on it), which also reduces the cookie size due to a smaller
    IV and less padding
  - use URI-safe base64 encoding to not waste space percent-encoding
    characters in the cookie
  - use . as the separator between cookie fields (instead of -, which
    is now a part of the base64 encoding)
  - if :expire_after is provided, add a timestamp to the cookie that's
    included in the HMAC to allow for secure expiration of cookies
    (preventing session replay attacks after the expiration of the
    cookie)
  - don't re-set the cookie on every request if the session hasn't
    changed (unless :expire_after is in use, in which case do it every
    5 minutes by default in order to refresh the timestamp)

ccutrer committed:
improvements and gemify!

 * don't bother encrypting the IV (?!)
 * use the same encryption key for the encryption as for the HMAC
   (easy to configure, since it's the same as for normal CookieStore)
 * compress the session before encrypting it (optional, will skip
   compression if it actually makes it larger)
 * use AES 128 CBC (AES 256 is actually *weaker*, due to a known
   attack on it), which also reduces the cookie size due to a smaller
   IV and less padding
 * use URI-safe base64 encoding to not waste space percent-encoding
   characters in the cookie
 * use . as the separator between cookie fields (instead of -, which
   is now a part of the base64 encoding)
 * if :expire_after is provided, add a timestamp to the cookie that's
   included in the HMAC to allow for secure expiration of cookies
   (preventing session replay attacks after the expiration of the
   cookie)
 * don't re-set the cookie on every request if the session hasn't
   changed (unless :expire_after is in use, in which case do it every
   5 minutes by default in order to refresh the timestamp)

FooBarWidget commented:
Very nice! But why not encrypt the IV? Is it supposed to be public information?

I'm opposed to removing the "It is prone to session replay attacks" sentence from the README. You are less prone now but not completely invulnerable. Attackers can still replay sessions within the timeout period.

ccutrer commented:
Because it's a waste of CPU cycles, adds unnecessary complexity, and does not add any "real" security (an IV is completely random data anyway; encrypting the randomness doesn't make it any more random, nor increase the security of bruteforcing the encrypted data): http://stackoverflow.com/questions/6167114/protect-encrypt-initialization-vector

I'll adjust the readme re: session replay attacks.

ccutrer committed:
Update README

Include more details on still being vulnerable to session replay attacks, just less vulnerable.

ccutrer committed:
fix bug with expiring cookie

test plan:
 * set :expire_after, but also set :expires to nil for session
   expiration of the real cookie
 * it shouldn't give a page error when the cookie expires

fourcolors committed:
Make options a class variable

Options is now a class variable that you can
change during call time. This way you can set
attributes on the options hash. In this case
we ensure the options hash gets the option
to set :expire_after

ccutrer committed:
Merge pull request #1 from fourcolors/master

Allow options to be changed.

ccutrer committed:
version 1.0.2

lukfugl committed:
record session_started

lukfugl committed:
rename session_started -> session_refreshed_at

the timestamp being used is refreshed every so often and doesn't
actually indicate the *start* of the session

ccutrer committed:
Merge pull request #2 from lukfugl/session_started

record session_refreshed_at

ccutrer committed:
bump to 1.0.3

lukfugl committed:
fix calculation of session_refreshed_at timestamp

the timestamp stored in the encrypted cookie already represents the
"session_refreshed_at" value. in 1.0.3 we incorrectly interpreted it as
a "expires_at" value, as if it already included the expire_after value

canadaduane committed:
Merge pull request #3 from lukfugl/session_started

fix calculation of session_refreshed_at timestamp

Commit:
bump to 1.0.4

Commit:
update gem date

maneframe committed:
update to work with rails 3.2

maneframe committed:
update date and authors

maneframe committed:
Merge pull request #4 from maneframe/rails3b

update to work with rails 3.2

maneframe committed:
allow setting expire_after before unmarshalling

lukfugl committed:
Merge pull request #6 from maneframe/expire_after

allow setting expire_after before unmarshalling

lukfugl committed:
bump version

maneframe committed:
use unsigned cookies in cookie_jar for rails 2 interoperability

ccutrer committed:
bump

ccutrer committed:
fix deprecation warnings in Ruby 2.1

maneframe committed:
Merge pull request #8 from ccutrer/master

fix deprecation warnings in Ruby 2.1

ccutrer committed:
bump version

ccutrer committed:
catch exceptions unmarshalling data

maneframe committed:
Merge pull request #9 from ccutrer/unmarshal_fix

catch exceptions unmarshalling data

ccutrer committed:
bump version

ccutrer committed:
allow rails 4.0 and 4.1

maneframe committed:
log unmarshalling errors

maneframe committed:
bump version

ccutrer committed:
Merge pull request #11 from ccutrer/rails4

allow rails 4.0 and 4.1

ccutrer committed:
bump version

ccutrer committed:
make load_session public

for rails4 compatibility

maneframe committed:
Merge pull request #12 from ccutrer/rails4

make load_session public

ccutrer committed:
bump version

ccutrer committed:
include the IV in the HMAC

and optionally still allow HMACs that didn't have it

ccutrer committed:
extract hmac into helper function

and fix bug where we passed iv twice

ccutrer committed:
Merge pull request #13 from ccutrer/iv_in_hmac

include the IV in the HMAC

ccutrer committed:
bump version

lukfugl committed:
remove accidental '@' in option key

ccutrer committed:
Merge pull request #14 from lukfugl/master

remove accidental '@' in option key

lukfugl committed:
bump version

maneframe committed:
fix destroy_session for rails 4

maneframe committed:
bump version

maneframe committed:
allow rails 4.2 and bump version

ccutrer committed:
upgrade rspec, lock to rails 4.2, fix spec

ccutrer committed:
drop legacy HMAC support

ccutrer committed:
remove shim for legacy OpenSSL

ccutrer committed:
refactor for newer rails' set_cookie/get_cookie methods

ccutrer committed:
don't use a unique env key for the cookie

then we don't need to override destroy_session at all

ccutrer committed:
assume rails 4+

ccutrer committed:
don't use a separate option key for per-request altering of expire_after

ccutrer committed:
rails 5/rack 2 support

ccutrer committed:
Merge pull request #15 from ccutrer/rails5

Rails 5

ccutrer committed:
bump version

ccutrer committed:
oops, still compatible with 4.2

ccutrer committed:
bump version

ccutrer committed:
fix enforcing expire_after

ccutrer committed:
bump version

ccutrer committed:
stop using deprecated constant

ccutrer committed:
bump version

ccutrer committed:
fix encryption for Ruby 2.4

Ruby 2.4 validates the encryption key length is exactly the
correct size, instead of just long enough. but be careful, because
we still use the secret for the HMAC, which can take a key of
any length. so only truncate what goes to encryption, not what
goes to HMAC

ccutrer committed:
Merge pull request #16 from ccutrer/keylength24

fix encryption for Ruby 2.4

ccutrer committed:
bump version

ccutrer committed:
allow rails 5.1

ccutrer committed:
bump version

maneframe committed:
rails 5.1: cast durations to integer before calculations

ccutrer committed:
fix refreshing the session on every request

keys are strings; don't use a symbol for the timestamp key

ccutrer committed:
Merge pull request #17 from ccutrer/string_key

fix refreshing the session on every request

ccutrer committed:
bump version

ccutrer committed:
bump version again to avoid conflict with released/uncommitted version

maneframe committed:
allow rails 5.2

maneframe committed:
bump version

ccutrer committed:
allow rails 6.0

ccutrer committed:
bump version

maneframe committed:
fix write_session to work with SessionId

https://github.com/rails/rails/commit/92ec9f270d696a17ba7604b3d02931f07c3e7c6a

refs #CNVS-49218

maneframe committed:
bump version

ccutrer committed:
allow rails 6.1

ccutrer committed:
Merge pull request #18 from ccutrer/rails61

allow rails 6.1

maths22 committed:
Allow rails 7.0

maths22 committed:
Merge pull request #19 from instructure/rails-70

Allow rails 7.0

maths22 committed:
Support rails 7.1

aogata-inst committed:
add support for rails 7.2

aogata-inst committed:
add support for rails 8.0

ccutrer committed:
support Rails 8.1 (#22)
